partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/DynamicTyper.kt

```diff
@@ -132,6 +146,47 @@ internal class DynamicTyper {
         return s to mapping
     }
 
+    private fun computeDecimal(): CompilerType? {
```

Comment:
Wonder if we should preserve the change from Yingtao's [comment](https://github.com/partiql/partiql-lang-kotlin/pull/1554#discussion_r1722419851) in commit [`873b6c4` (#1554)](https://github.com/partiql/partiql-lang-kotlin/pull/1554/commits/873b6c4f4ce835ded2dcc91e091cf1895c8b0f52). Basically,
- Make it clearer that this function is only for computing constrained decimals. So rename to `computeConstrainedDecimal`.
- Error when we receive an arbitrary decimal rather than returning `null`, since that would indicate a bug in the dynamic typer. This function shouldn't be called when we have previously encountered an arbitrary decimal.

Reply:
The merged change is still wrong. The supertype of INT, SMALLINT, and DECIMAL(5, 0), according to #1554 is DECIMAL. Though, when `computeDecimal` is invoked to compute the precision and scale, it throws an error when folding over the types of INT and SMALLINT.

This PR merges the not completely correct code while mitigating the thrown exception. I tried to apply as minimal changes as possible to the merge, but this is one area known to have a bug.

Reply:
Ok -- then if the merged behavior is misaligned w/ the change in 0.14.8, then we should either
1. track it in a TODO and tracking issue
2. or fix it in this PR if it's not too much work

Reply:
https://github.com/partiql/partiql-lang-kotlin/issues/1566